File: 2024/python/day4.py
def part1() -> int:
    grid = []
    with open("../data/input4.txt") as inp:
        for line in inp.readlines():
            grid.append(list(line.strip()))

    count = 0

    valid_coords = set((i, j) for i in range(len(grid)) for j in range(len(grid[0])))

    for x in range(len(grid)):
        for y in range(len(grid[0])):
            directions = [
                (i, j) for i in [-1, 0, 1] for j in [-1, 0, 1] if (i, j) != (0, 0)
            ]
            for dx, dy in directions:
                if (x + dx * 3, y + dy * 3) not in valid_coords:
                    continue

                if [grid[x + dx * n][y + dy * n] for n in range(4)] == list("XMAS"):
                    count += 1

    return count


def part2() -> int:
    grid = []
    with open("../data/input4.txt") as inp:
        for line in inp.readlines():
            grid.append(list(line.strip()))

    count = 0

    valid_coords = set((i, j) for i in range(len(grid)) for j in range(len(grid[0])))

    for x in range(len(grid)):
        for y in range(len(grid[0])):
            if (x + 2, y + 2) not in valid_coords:
                continue

            if grid[x + 1][y + 1] != "A":
                continue

            if [grid[x + n][y + n] for n in range(3)] in [
                list("MAS"),
                list("SAM"),
            ] and [grid[x + n][y + 2 - n] for n in range(3)] in [
                list("MAS"),
                list("SAM"),
            ]:
                count += 1

    return count


print(part1())
print(part2())

File: 2024/rust/src/day4.rs
use std::collections::HashSet;

use crate::day::{Day, Solution};

pub struct Day4 {
    day: Day,
}

impl Day4 {
    pub fn new() -> Self {
        Self { day: Day::new(4) }
    }
}

impl Solution for Day4 {
    fn part1(&self) -> u64 {
        let input = include_str!("../../data/input4.txt");
        let grid: Vec<Vec<char>> = input.lines().map(|line| line.chars().collect()).collect();

        let directions: Vec<(i32, i32)> = (-1..=1)
            .flat_map(|dx| (-1..=1).map(move |dy| (dx, dy)))
            .filter(|&(dx, dy)| dx != 0 || dy != 0)
            .collect();

        let n = grid.len();
        let mut count = 0;

        let mut valid_coords: HashSet<(i32, i32)> = HashSet::with_capacity(n * 2);
        for m in 0..n {
            for n in 0..n {
                valid_coords.insert((m as i32, n as i32));
            }
        }

        for x in 0..n {
            for y in 0..n {
                for (dx, dy) in &directions {
                    if !valid_coords.contains(&(x as i32 + dx * 3, y as i32 + dy * 3)) {
                        continue;
                    }

                    let word: String = (0..4)
                        .map(|n| grid[x + *dx as usize * n][y + *dy as usize * n])
                        .collect();

                    if word == "XMAS" {
                        count += 1;
                    }
                }
            }
        }

        count
    }

    fn part2(&self) -> u64 {
        let input = include_str!("../../data/input4.txt");
        let grid: Vec<Vec<char>> = input.lines().map(|line| line.chars().collect()).collect();

        let n = grid.len();
        let mut count = 0;

        let mut valid_coords: HashSet<(i32, i32)> = HashSet::with_capacity(n * 2);
        for m in 0..n {
            for n in 0..n {
                valid_coords.insert((m as i32, n as i32));
            }
        }

        for x in 0..n {
            for y in 0..n {
                if !valid_coords.contains(&(x as i32 + 2, y as i32 + 2)) {
                    continue;
                }

                if grid[x + 1][y + 1] != 'A' {
                    continue;
                }

                let word1: String = (0..3).map(|n| grid[x + n][y + n]).collect();
                let word2: String = (0..3).map(|n| grid[x + n][y + 2 - n]).collect();

                let allowed_words = ("MAS", "SAM");

                if (word1 == allowed_words.0 || word1 == allowed_words.1)
                    && (word2 == allowed_words.0 || word2 == allowed_words.1)
                {
                    count += 1;
                }
            }
        }

        count
    }

    fn get_day(&self) -> u8 {
        self.day.day
    }
}

pub fn run() {
    Day4::new().solve();
}
